src/mito2/src/memtable/bulk/part.rs

```
@@ -209,6 +217,284 @@ impl BulkPart {
     }
 }
 
+/// Builder type for primary key dictionary array.
+type PrimaryKeyArrayBuilder = BinaryDictionaryBuilder<UInt32Type>;
+
+/// Primary key column builder for handling strings specially.
+enum PrimaryKeyColumnBuilder {
+    /// String dictionary builder for string types.
+    StringDict(StringDictionaryBuilder<UInt32Type>),
+    /// Generic mutable vector for other types.
+    Vector(Box<dyn MutableVector>),
+}
+
+impl PrimaryKeyColumnBuilder {
+    /// Appends a value to the builder.
+    fn push_value_ref(&mut self, value: ValueRef) -> Result<()> {
+        match self {
+            PrimaryKeyColumnBuilder::StringDict(builder) => {
+                if let Ok(Some(s)) = value.as_string() {
+                    // We know the value is a string.
+                    builder.append_value(s);
+                } else {
+                    builder.append_null();
```

is fail instead of push null here a better option to discover bugs early?